wrapanapi/systems/google.py

```diff
@@ -0,0 +1,798 @@
+# Mimetype to use if one can't be guessed from the file extension.
+DEFAULT_MIMETYPE = 'application/octet-stream'
+
+# List of image projects which gce provided from the box. Could be extend in the future and
+# will have impact on total number of templates/images
+IMAGE_PROJECTS = ['centos-cloud', 'debian-cloud', 'rhel-cloud', 'suse-cloud', 'ubuntu-os-cloud',
+                  'windows-cloud', 'opensuse-cloud', 'coreos-cloud', 'google-containers']
+
+
+class GCEInstance(Instance):
+    state_map = {
+        'PROVISIONING': VmState.STARTING,
+        'STAGING': VmState.STARTING,
+        'STOPPING': VmState.STOPPING,
+        'RUNNING': VmState.RUNNING,
+        'TERMINATED': VmState.STOPPED,
+    }
+
+    def __init__(self, system, raw=None, **kwargs):
+        """
+        Constructor for GCEInstance
+
+        Args:
+            system: GCEInstance object
+            raw: the raw json data for this instance returned by the compute API
+            name: the name of the VM
+            zone: the zone of the VM
+        """
+        self._name = raw['name'] if raw else kwargs.get('name')
+        self._zone = raw['zone'].split('/')[-1] if raw else kwargs.get('zone')
+        if not self._name or not self._zone:
+            raise ValueError("missing required kwargs: 'name' and 'zone'")
+
+        super(GCEInstance, self).__init__(system, raw, **kwargs)
+
+        self._project = self.system._project
+        self._api = self.system._compute.instances()
+
+    @property
+    def _identifying_attrs(self):
+        return {'name': self._name, 'zone': self._zone, 'project': self._project}
+
+    @property
+    def uuid(self):
+        return self.raw['id']
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def zone(self):
+        return self._zone
+
+    def refresh(self):
+        try:
+            self.raw = self._api.get(
+                project=self._project, zone=self._zone, instance=self._name).execute()
+        except errors.HttpError as error:
+            if error.resp.status == 404:
+                raise VMInstanceNotFound(self._name)
+            else:
+                raise
+        return self.raw
+
+    def _get_state(self):
+        self.refresh()
+        return self._api_state_to_vmstate(self.raw['status'])
+
+    @property
+    def ip(self):
+        self.refresh()
+        return self.raw.get('networkInterfaces')[0].get('networkIP')
+
+    @property
+    def ip_external(self):
+        self.refresh()
+        access_configs = self.raw.get('networkInterfaces')[0].get('accessConfigs')[0]
+        return access_configs.get('natIP')
+
+    @property
+    def type(self):
+        if self.raw.get('machineType', None):
+            return self.raw['machineType'].split('/')[-1]
+        return None
+
+    @property
+    def creation_time(self):
+        creation_time = iso8601.parse_date(self.raw['creationTimestamp'])
+        return creation_time.astimezone(pytz.UTC)
```

Comment:
thanks for keeping this part consistent, our VM cleanup is based on comparing these timestamps :)